compiler/damlc/daml-compiler/src/DA/Daml/Compiler/DataDependencies.hs

```diff
@@ -467,48 +523,53 @@ isConstraint = \case
 genModule :: Env -> LF.PackageRef -> LF.ModuleName -> Gen Module
 genModule env pkgRef modName = do
     let Config{..} = envConfig env
-        isStable
-            | LF.PRSelf <- pkgRef = False
-            | LF.PRImport pkgId <- pkgRef =
-                pkgId `Set.member` configStablePackages
+        (packageQualified, modName') = case pkgRef of
+            LF.PRImport pkgId
+                | pkgId `Set.member` configDependencyPackages -> (False, prefixDependencyModule pkgId modName)
+                | pkgId `Set.member` configStablePackages -> (False, prefixModuleName configSdkPrefix modName)
+            _ -> (True, modName)
         unitId = configGetUnitId pkgRef
-    genModuleAux env isStable unitId modName
+    genModuleAux packageQualified unitId modName'
 
 genStableModule :: Env -> UnitId -> LF.ModuleName -> Gen Module
-genStableModule env = genModuleAux env True
-
-genModuleAux :: Env -> Bool -> UnitId -> LF.ModuleName -> Gen Module
-genModuleAux env isStable unitId modName = do
-    let sdkPrefix = configSdkPrefix (envConfig env)
-        newModName
-            | isStable = LF.ModuleName (sdkPrefix ++ LF.unModuleName modName)
-            | otherwise = modName
-        ghcModName = mkModuleName . T.unpack $ LF.moduleNameString newModName
-        modRef = ModRef isStable unitId newModName
+genStableModule env unitId = genModuleAux False unitId . prefixModuleName (configSdkPrefix $ envConfig env)
+
+prefixModuleName :: [T.Text] -> LF.ModuleName -> LF.ModuleName
+prefixModuleName prefix (LF.ModuleName mod) = LF.ModuleName (prefix <> mod)
+
+prefixDependencyModule :: LF.PackageId -> LF.ModuleName -> LF.ModuleName
+prefixDependencyModule (LF.PackageId pkgId) = prefixModuleName ["Pkg_" <> pkgId]
+
+genModuleAux :: Bool -> UnitId -> LF.ModuleName -> Gen Module
+genModuleAux isQualified unitId modName = do
+    let ghcModName = mkModuleName . T.unpack $ LF.moduleNameString modName
+        modRef = ModRef isQualified unitId modName
     emitModRef modRef
     pure $ mkModule unitId ghcModName
 
-convType :: Env -> LF.Type -> Gen (HsType GhcPs)
-convType env =
+convType :: Env -> MS.Map LF.TypeSynName LF.PackageId -> LF.Type -> Gen (HsType GhcPs)
+convType env reexported =
     \case
         LF.TVar tyVarName -> pure $
             HsTyVar noExt NotPromoted $ mkRdrName $ LF.unTypeVarName tyVarName
 
         ty1 LF.:-> ty2 -> do
-            ty1' <- convType env ty1
-            ty2' <- convType env ty2
+            ty1' <- convType env reexported ty1
+            ty2' <- convType env reexported ty2
             pure $ if isConstraint ty1
                 then HsQualTy noExt (noLoc [noLoc ty1']) (noLoc ty2')
                 else HsParTy noExt (noLoc $ HsFunTy noExt (noLoc ty1') (noLoc ty2'))
 
         LF.TSynApp LF.Qualified{..} lfArgs -> do
-            ghcMod <- genModule env qualPackage qualModule
+            let pkg | Just pkgId <- MS.lookup qualObject reexported = LF.PRImport pkgId
```

Comment:
1) Is this necessary? Shouldn't the fact that the typeclass is being re-exported explicitly, allow us to use the fully qualified name of the module that re-exports it, instead of module name being re-exported? (This is why we re-export typeclasses in the first place.)

2) If it is necessary, won't this check conflate two typeclasses with the same name? One may be re-exported and the other not. Granted, this is an edge case, but perhaps we should be looking up by qualModule as well.

Reply:
1. It is necessary for typeclasses reexported from the current module (those are the only ones in `reexported`). You cannot refer to them qualified using the current module name. I’ve added a comment and factored it out into a separate definition. Does that sounds sensible?

2. Good point, I’ve added more checks.

Reply:
👍 Ok, I suspected that might be the case, but I really wasn't sure. That also explains why it wasn't an issue in the tests I wrote (they didn't test same module reexports). The `rewriteClassReexport` looks good.